### PR TITLE
Add prometheus metric exports, tests.

### PR DESCRIPTION
**Pull request by gfr10598**
Adds metrics for prometheus, with unit tests.
Adds test for logConnection.

---
This change is [Reviewable](https://reviewable.io/reviews/npad/sidestream/24)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,12 +8,12 @@ before_install:
...
-  - pip install -r requirements.txt
-  - pip install -r test-requirements.txt
-  - sudo pip install coverage  # Because we have to call if from sudo
+  # Install dependencies with sudo, so that they can be used from sudo
+  - sudo pip install -r requirements.txt
+  - sudo pip install -r test-requirements.txt
 
 script:
-  - sudo python -m coverage run --source='.' --omit '*_test.py' -m unittest discover --pattern='*_test.py'
+  - sudo coverage run --source='.' --omit '*_test.py' -m unittest discover --pattern='*_test.py' && echo $?
```
Comment (pboothe) [inline .travis.yml]: debug `&& echo $?` left in

**Comment by stephen-soltesz** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[exitstats.py, line 49 at r2](https://reviewable.io:443/reviews/npad/sidestream/24#-KbHLzgfYI9E4gehWU8U:-KbHLzggvvZLeS47_VCH:b-4zbw8q) ([raw file](https://github.com/npad/sidestream/blob/716fda9d8b5bac79f2464a7faf536e5c96c4ba25/exitstats.py#L49)):*
> ```Python
> 
> PROMETHEUS_SERVER_PORT = 9090
> connection_count = prom.Counter('connection_count',
> ```

The Prometheus docs have some helpful notes on best practices notes for metric and label naming.

https://prometheus.io/docs/practices/naming/#metric-names
https://prometheus.io/docs/practices/instrumentation/#use-labels

In particular, in this case, it could be helpful to prefix these counters with an application name, such as "sidestream_".

---

*[exitstats.py, line 179 at r2](https://reviewable.io:443/reviews/npad/sidestream/24#-KbHRoj1aFKFg4mSs4Y1:-KbHRoj1aFKFg4mSs4Y2:b-9ha2ug) ([raw file](https://github.com/npad/sidestream/blob/716fda9d8b5bac79f2464a7faf536e5c96c4ba25/exitstats.py#L179)):*
> ```Python
>     return
>   if snap["RemAddress"].startswith("128.112.139"):
>     plc_connection_count.inc()
> ```

Here, since `conection_count` and `plc_connection_count` are mutually exclusive and of the same type of thing, it could be helpful to use labels to distinguish them on a single metric. For example, something like:

    sidestream_connection_count{remote="plc"}
    sidestream_connection_count{remote="public"}

I'm not loving those names, but that's the idea.

Then we could run queries on the resulting data like:

    topk(10, sum(sidestream_connection_count) by (hostname) keep_common))

To find the top 10 machines with the most of all kinds of connections. (We could run such a query with separate metric names, but the expression becomes a little more complex, and some of the values could be forgotten).

---

*[exitstats.py, line 213 at r2](https://reviewable.io:443/reviews/npad/sidestream/24#-KbHRawbwuJIQgdhrPjd:-KbHRawciv5d3pKO_vy6:byu06al) ([raw file](https://github.com/npad/sidestream/blob/716fda9d8b5bac79f2464a7faf536e5c96c4ba25/exitstats.py#L213)):*
> ```Python
>   closed = []
>   while True:
>     main_loop_count.inc()
> ```

Do we need this counter?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/npad/sidestream/24)*

File: exitstats_test.py [hunk 2]
```diff
@@ -9,10 +9,59 @@
...
+    result = FakeConnection()
+    result.cid = self.cid
+    result.values = self.values
+    return result
+
+  def __iter__(self):
+    return values.__iter__()
+
+class TestMonitoring(unittest.TestCase):
+  def setUp(self):
```
Comment (pboothe) [inline exitstats_test.py]: Do you need a `tearDown` to kill the server?

File: exitstats_test.py [hunk 3]
```diff
@@ -9,10 +9,59 @@
...
+
+    # Read from the httpserver and assert the correct connection count.
+    url = "http://localhost:%s"%(exitstats.PROMETHEUS_SERVER_PORT)
+    # Read in a while loop, since the server is a daemon and may not start immediately.
+    while True:
+      try:
+        response = urllib2.urlopen(url).read()
+        print("ok")
+        break;
+      except urllib2.URLError as e:
```
Comment (pboothe) [inline exitstats_test.py]: Can you turn this into a bounded loop? If it hasn't started in 1000 tries it probably won't start in 1001.

I recommend the little-used 
```python
for _ in xrange(1000):
  do stuff and possibly break
else:
  exited the loop without break, raise an error
```
syntax

File: exitstats_test.py (hunk 2 above)
Comment (gfr10598) [inline exitstats_test.py]: No documentation on how to shut down the server, so just adding a comment.

File: .travis.yml (hunk 1 above)
Comment (gfr10598) [inline .travis.yml]: Removed.  Still not quite sure why it wasn't working without it, but it seems ok now.

File: exitstats_test.py (hunk 3 above)
Comment (gfr10598) [inline exitstats_test.py]: Done

**Comment by gfr10598** [comment]
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[exitstats.py, line 49 at r2](https://reviewable.io:443/reviews/npad/sidestream/24#-KbHLzgfYI9E4gehWU8U:-KbHsYlt22Y8lrBUuEO2:b-896fix) ([raw file](https://github.com/npad/sidestream/blob/716fda9d8b5bac79f2464a7faf536e5c96c4ba25/exitstats.py#L49)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

The Prometheus docs have some helpful notes on best practices notes for metric and label naming.

https://prometheus.io/docs/practices/naming/#metric-names
https://prometheus.io/docs/practices/instrumentation/#use-labels

In particular, in this case, it could be helpful to prefix these counters with an application name, such as "sidestream_".
</blockquote></details>

Done.

---

*[exitstats.py, line 179 at r2](https://reviewable.io:443/reviews/npad/sidestream/24#-KbHRoj1aFKFg4mSs4Y1:-KbHsft9uzB3Gn1yYloB:b-896fix) ([raw file](https://github.com/npad/sidestream/blob/716fda9d8b5bac79f2464a7faf536e5c96c4ba25/exitstats.py#L179)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Here, since `conection_count` and `plc_connection_count` are mutually exclusive and of the same type of thing, it could be helpful to use labels to distinguish them on a single metric. For example, something like:

    sidestream_connection_count{remote="plc"}
    sidestream_connection_count{remote="public"}

I'm not loving those names, but that's the idea.

Then we could run queries on the resulting data like:

    topk(10, sum(sidestream_connection_count) by (hostname) keep_common))

To find the top 10 machines with the most of all kinds of connections. (We could run such a query with separate metric names, but the expression becomes a little more complex, and some of the values could be forgotten).
</blockquote></details>

Done.

---

*[exitstats.py, line 213 at r2](https://reviewable.io:443/reviews/npad/sidestream/24#-KbHRawbwuJIQgdhrPjd:-KbHsqoSw67iekZIM4tw:b-896fix) ([raw file](https://github.com/npad/sidestream/blob/716fda9d8b5bac79f2464a7faf536e5c96c4ba25/exitstats.py#L213)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Do we need this counter?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/npad/sidestream/24)*

**Pull request merged**